### DISCUSSION TITLE
Async required method signature

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5957] [Language Ideas]
It's a question that comes up from time to time: currently, the only thing in the method signature pertaining to asynchronous programming is the return type: if it's task-like, it can be awaited. However, the task can be stored and awaited later. This can cause unexpected behavior with exceptions, if the method implementation happens to have `async` with state machine (which eats the exception and throws on `await`), vs if it's a synchronous method which returns `Task.CompletedTask` or something like that. This introduces a situation where the method signature does not guarantee the method's behavior.
```cs
void DoSomethingElse() => throw new IOException();
public Task DoSomethingAsync()
{
    DoSomethingElse();
    return Task.CompletedTask;
}

var doSomethingTask = DoSomethingAsync(); // Program crashes here because DoSomethingAsync didn't use `async`
// Do something else
try
{
    await doSomethingTask;
}
catch (IOException)
{
    // This catch block can only catch exceptions that were returned in a Task value from DoSomethingAsync
}

// this fixes the exception problem, but shows a warning - 
// because the language thinks you are creating a state machine for no reason
public async Task DoSomethingAsync2()
{
    DoSomethingElse();
}
```
To solve this, I propose to add a new return type modifier, "async required". This means that the method is guaranteed to have a state machine and only throw on await.
```cs
void DoSomethingElse() => throw new IOException();
public async required Task DoSomethingAsync() // no warning; we told the compiler that we're promising "async"
{
    DoSomethingElse();
}

var doSomethingTask = DoSomethingAsync(); // this line never throws, guaranteed (other than things like StackOverflow/OutOfMemory)
// Do something else
try
{
    await doSomethingTask;
}
catch (IOException)
{
    // we caught the exception correctly
}
```
That return type modifier can be applied to interface methods, delegates, and actual methods.

**Update:** when compiling `DoSomethingAsync`, C# can deduce that since it's `async`, and no `await` calls are made, the result can be simplified to:
```cs
Task DoSomethingAsync()
{
    try
    {
        DoSomethingElse();
        return Task.CompletedTask;
    }
    catch(Exception e)
    {
        return Task.FromException(e);
    }
}
```
This way, we get the best of both worlds: the caller gets a guarantee that only `await` needs to be in `try-catch`, and much less overhead is generated when the method is actually synchronous.

**Comment by bernd5** [comment]
It is possible to write:

[Sample:](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyBWAbmrsbIDoBJAeXZUOKegBEIAZQgBbGBgAWASzwBzAKIAbAM4wAFAEp6AXgB89RVAgB3enhi2B6hGBgAHDEoh5DQoknEpWXllNQBBbQBPPDBDegB6eLsIemsAQyg8FVVWVJgLCE0AEwsFfMgZNyVNGChStIw8gHJYejcrGSVtbPoAIiIADl7qAG9qegmLKEjxybGqScXA6TlFbK1dX1mlxgB2RiRuAGFZNxqMGCL/IUWAX236MAawBT1nVw8vPHoYAwf5nYTIj7fzcABiHXe7k83j0vxuk3uVCR1AAbhl6EUgqtQqp/CZlsE1uEojFfAkkmttPRNCp8vZUbVSlZrNo0PRVDgMmk8BdLvQ9BBFEzFDzStlqbSANb5SQYNJgKX8RlQABmmhs8X4OAw/FVAFkYDJoJE/lREoF6NoibifjoYNQMNNRg8iABOTHYkLZa7UJFPDAvAVOFzQr5mgGTC3WcppHCqBSNYU/UOfbyPaCwMAYTQzBaTJhuvS9AMvbjl3oGIS3IA===)
```csharp
using System;
using System.Threading.Tasks;
using System.IO;

void DoSomethingElse() => throw new IOException();
Task DoSomethingAsync() // no warning; we told the compiler that we're promising "async"
{
    try
    {
        DoSomethingElse();
        return Task.CompletedTask;
    }
    catch(Exception e)
    {
        return Task.FromException(e);
    }
}

var doSomethingTask = DoSomethingAsync(); // this line never throws, guaranteed (other than things like StackOverflow/OutOfMemory)
// Do something else
try
{
    await doSomethingTask;
}
catch (IOException)
{
    // we caught the exception correctly
    Console.WriteLine("catch...");
}
```

This throws on await - not on task construction (consumer perspective).
In addition this approach avoids the allocation of Tasks and uses the shared `Task.CompletedTask` which makes it faster (less allocations).

**Reply by TahirAhmadov** [reply]
You're right - that's a good idea if you control both sides of the equation. Meaning, "this line never throws, guaranteed" is only true if you guarantee it to yourself.
My idea with adding a modifier to the signature is to allow this to be safe across libraries/projects/teams.

**Reply by bernd5** [reply]
personally I don't think that it is wrong to throw on task acquisition...

**Reply by TahirAhmadov** [reply]
In some cases, it's not; in others, it is. I can imagine if you want to parallelize many tasks, and do not want to try-catch the acquisition of each task, and instead just try-catch around Task.WhenAll, or whatever.
I mean, the existing syntax continues working as is; this would be a purely new thing for those situations where it's needed (where throwing on task acquisition is bad).

**Reply by bernd5** [reply]
Adding a keyword is quite expensive for such a thing...
What is a real world use case where an exception can occure in combination with `Task.CompletedTask`? For me this is more a theoretical concern...

**Reply by TahirAhmadov** [reply]
It is; I posted this just so that the conversation I had in a different thread wasn't forgotten.
The use case is:
```cs
interface IReader { Task<byte> Read(); }
class SocketReader: IReader { async Task<byte> Read() { ... } }
class MemoryReader:IReader 
{
  Task<byte> Read()=>Task.FromResult(this._buffer[this._index++]); // this can throw, if _buffer is null or _index somehow went out of range
}
IReader reader = GetReader(); // we don't know if it's socket or memory
var readTask = reader.Read(); // if it's memory, this can throw - program crash
// ... other code ...
try
{
  await readTask;
}
catch(Exception ex){ ... } // if it's socket, we catch correctly
```

**Reply by jnm2** [reply]
Any time you access a property or do really anything "extra" when delegating to another Task-returning method.

**Reply by TahirAhmadov** [reply]
Actually, @bernd5, your snippet is pertinent. C# compiler can look at `async required` modifier; look at the method body; it knows no `await` calls are present; and it can simplify from full-blown state machine, to basically your version with returning `Task.FromException(e)`. Bingo.

**Comment by svick** [comment]
To me, this seems unnecessary. C# does not attempt to define when exceptions can be thrown. I don't think this should be the one place in the language where it does.

**Reply by TahirAhmadov** [reply]
That's the point - `async` methods are significantly different from regular sync methods, and therefore warrant a special syntax.

**Reply by svick** [reply]
How are they different? For example, [property getters generally shouldn't throw exceptions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/property?WT.mc_id=DOP-MVP-5002735) either, but there is nothing in the language enforcing that.

**Reply by HaloFour** [reply]
Whether or not a method is annotated with `async` is entirely an implementation detail.  It is expected to be able to convert a method from `async` to non-`async` without any consumer code being aware of the change.  Nobody should assume that a method suffixed with `Async` will never throw, because that's not true.

**Reply by TahirAhmadov** [reply]
Properties are irrelevant, because they don't have 2 completion points, like task-returning methods do.
Have you read the OP? Current syntax either forces the caller to try-catch both task acquisition and awaiting, or forces all task-returning methods to be `async`. Both of those options are worse than having an explicit syntax for it.

**Reply by TahirAhmadov** [reply]
> Whether or not a method is annotated with `async` is entirely an implementation detail. It is expected to be able to convert a method from `async` to non-`async` without any consumer code being aware of the change.

That doesn't change. That's why my proposal calls for a new keyword, something like `async required`, which **is** part of the signature.

>  Nobody should assume that a method suffixed with Async will never throw, because that's not true.

Yes, and with the new `async required` modified, it guarantees that it will never throw, and only `await` should be try-caught.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> That doesn't change. That's why my proposal calls for a new keyword, something like `async required`, which is part of the signature.

Making an implementation detail part of the signature seems highly undesirable.  And that implementation detail says nothing about the method not being able to throw.  For example, `!!` in an `async` method will throw immediately, not return a failed `Task`, and that is by design.

**Reply by TahirAhmadov** [reply]
> Making an implementation detail part of the signature seems highly undesirable.

It is desirable if you don't want to write two try-catch statements each time you want to call a task-returning method. And, I want to stress the point that it's a new keyword specifically to _change it_ from an implementation detail to a method signature. Think of it like this: 
```cs
object Add(object left, object right) { if (left is int leftInt && right is int rightInt) { return leftInt + rightInt; } else throw new ArgumentException(); }
// now we change it to
int Add(int left, int right) { return left + right; }
```
It's the same idea: previously, the method could add numbers, but the signature said nothing about the types - it just accepted `object`s. Then, we decided to change parameters to `int`, while keeping the code the same, for type safety. Somebody can say, hey, why did you expose an implementation detail as a signature? And that question would be wrong. Do you agree?

>  For example, !! in an async method will throw immediately, not return a failed Task, and that is by design.

That's currently the case for `async` methods, but what's stopping C# from adopting a different rule for `async required` methods? Again, this would be a new feature - existing `async` stuff does not change one bit.

**Reply by svick** [reply]
@TahirAhmadov I don't see a meaningful distinction in this regard between `x.Property.Method()` and `await x.MethodAsync()`. In both cases, you have two places that could throw an exception and you might not except the first place to throw. Why is it important to ensure that `x.MethodAsync()` doesn't throw, but it's fine if `x.Property` throws?

> Current syntax either forces the caller to try-catch both task acquisition and awaiting, or forces all task-returning methods to be `async`.

I don't see having to extend the `try` to cover both task acquisition and awaiting as a problem that necessitates new syntax. Also, there is a third option: understanding when exceptions can be thrown. Though for framework methods, that would require better documentation (see https://github.com/dotnet/dotnet-api-docs/issues/7840).

**Reply by TahirAhmadov** [reply]
>  For example, !! in an async method will throw immediately, not return a failed Task, and that is by design.

Leaving aside the wisdom of making that decision, and whether it's possible or it's too late to change it for regular `async` methods, the fact that the decision was made to null-check immediately actually adds a very strong argument for my proposal above: namely, one of the ways some people used to avoid the exception problem - making all task-returning methods `async` - does not guarantee that only `await` throws. If anything, this makes it even more important to introduce `async required`.

**Reply by TahirAhmadov** [reply]
> @TahirAhmadov I don't see a meaningful distinction in this regard between `x.Property.Method()` and `await x.MethodAsync()`. 

There is a huge distinction there - `x.Property.Method()` is two method invocations, and `await x.MethodAsync()` is one method invocation, from language perspective. And the implementation detail of how `MethodAsync` does the state machine and what not is irrelevant to the language concepts in play.

> > Current syntax either forces the caller to try-catch both task acquisition and awaiting, or forces all task-returning methods to be `async`.
> 
> I don't see having to extend the `try` to cover both task acquisition and awaiting as a problem that necessitates new syntax. 

Extending it is possible in some cases, but not in others; what if you have a few methods which acquire tasks, and then some other method parallelizes them and does `await Task.WhenAll`? And checking for exceptions every time is needless copy-pasta, which is error-prone and takes up valuable time. 

> Also, there is a third option: understanding when exceptions can be thrown. Though for framework methods, that would require better documentation (see [dotnet/dotnet-api-docs#7840](https://github.com/dotnet/dotnet-api-docs/issues/7840)).

That's talking about documentation, and while I'm all for documentation, it doesn't stop bad code from compiling - which is my objective here.

**Reply by HaloFour** [reply]
It's not bad code for an async method to throw immediately, though.  Validation code in particular should check and throw immediately, there's no purpose in wrapping that within a `Task`.  That's why it makes perfect sense for `!!` to throw and not to return a failed `Task`.

**Reply by TahirAhmadov** [reply]
> It's not bad code for an async method to throw immediately, though. Validation code in particular should check and throw immediately, there's no purpose in wrapping that within a `Task`. That's why it makes perfect sense for `!!` to throw and not to return a failed `Task`.

Define "bad". Does having to repeat your try-catch statements every time make it bad? From DRY perspective, it is bad. And I would argue that DRY is one of the most important things in software.

**Reply by HaloFour** [reply]
This has nothing to do with DRY.  Methods and Tasks have separate failure modes.

**Reply by TahirAhmadov** [reply]
Why should they, though?

**Reply by HaloFour** [reply]
Exceptions resulting from validation of method parameters should be kept as closely associated with the source as possible.  They are not part of the asynchronous operation represented by that task, nor is it helpful to observe that exception far from where it originated, which is pertinent if you are passing the task around instead of awaiting it immediately.

**Reply by jnm2** [reply]
Jon Skeet makes this distinction and calls them "boneheaded exceptions": exceptions which indicate a bug in the programming as opposed to a reality that you might theoretically  intend to program a mitigation for, such as an I/O exception. Since you aren't catching boneheaded exceptions, except possibly at a high-level agnostic exception handler that is agnostic to this particular call, it's okay for the method to throw them instead of returning a task object wrapping them.
